proposals/0000-reduce-fragmentation.md

```diff
@@ -0,0 +1,332 @@
+    backgroundColor: 'hsla(50,50,50,0.5)',
+    pointerEvents: 'none',
+    transform: 'scale(0.9)',
+    width: '5rem'
+  }}
+>
+
+<Text
+  style={{
+    lineClamp: 3,
+    userSelect: 'none',
+    verticalAlign: 'middle'
+  }}
+>
+
+<Image
+  style={{
+    objectFit: 'cover'
+  }}
+>
+```
+
+### New imperative APIs
+
+* `EventTarget` API
+  * The `EventTarget` API is a implemented in JS environments, and should be available on React Native host elements. This feature is commonly used on web, and could be used by React Native developers to support more complex features without first requiring further core API changes to support each use case.
+  * Promote `node.unstable_addEventListener`, `node.unstable_removeEventListener` to stable APIs.
+  * Add `node.dispatchEvent` support.
+  * This cannot be shimmed in user-space.
+* `node.getBoundingClientRect`
+  * Synchronous element measurement API. Current APIs like `node.measure` are async and return slightly different information to what you expect from `getBoundingClientRect` on web.
+  * This cannot be shimmed in user-space.
+
+### Other
+
+* Revisit the type of all events modeled on W3C events. For example, `onTouchStart` should receive an event object that conforms to that received by the same prop in React DOM.
+* Aim to deprecate `StyleSheet.flatten()`.
+  * This API encourages runtime introspection of styles in render calls. This pattern is a performance cost (flattening arrays of objects), and prevents us from supporting build-time optimizations for web (extracting styles to CSS files).
+* Aim to deprecate `StyleSheet.compose()`.
+  * This API has no real value.
+* Aim to restore obfuscation of styles passed to `StyleSheet.create()`.
+  * Returning the styles from create calls encourages introspection of those styles. On web, we need to be able to remove the JavaScript style objects from bundles to support build-time optimization like extraction to CSS files.
+* Aim to deprecate the `Touchable*` components.
+  * The `Pressable` component was shipped as a replacement for `Touchable*`.
+* Aim to support more interaction states with `Pressable`.
+  * Add ‘hovered’ and ‘focused’ to the state object, and replace ‘testOnly_pressed’ with an object of states.
+* Aim to move `Animated` out of core.
+  * Move Animated out of core to separate package in monorepo, allow React Native for Web to import this package. Eventually could drop Animated from ‘react-native’ export.
+* Aim to move `VirtualizedList` and `FlatList` out of core.
+  * Move `VirtualizedList`/`FlatList` out of core to separate package in monorepo, allow React Native for Web to import this package. Eventually could drop these components from `'react-native'` export.
+* Compile to standard JavaScript before publishing to npm (make this the default for the community). React Native and its community packages currently publish uncompiled, flow-typed JavaScript to npm. This causes common tools to crash on these packages.
```

Comment:
I'd love to understand why this wasn't done years ago. Metro would run a lot faster if it didn't have to transpile everything all the time. 

This one feature might warrant its own proposal given the scope. Perhaps @motiz88 also has some thoughts on this.

Reply:
Yes, I second making this its own proposal - the devil's in the details here.

1. If the entire ecosystem switched to publishing standard JavaScript to npm overnight (which realistically would be a complicated migration in its own right) - and if that in practice means compiling to the lowest common denominator - there's a real risk of bloating bundle sizes and losing out on runtime performance because of poorly/inconsistently transpiled packages. (Think packages bringing their own incompatible versions of Babel helpers, or transpiling features that the target runtime actually supports natively.) Where now we have leverage to compile things in the best way for each app, we would spread that leverage across a long tail of community-maintained packages. So IMO we'd need to give clear guidance (+ recommend specific tools/configs?) to package authors re: what and how much to transpile, so we can still influence the overall outcome for app authors.
2. By "standard JavaScript" are we also thinking of moving away from React Native-isms like platform-specific file extensions, asset scale file extensions (`@2x.png` etc), and reliance on spec-breaking compiler features such as [inline requires](https://reactnative.dev/docs/ram-bundles-inline-requires#inline-requires) and `Platform.select()` inlining? IMO these are good to explore in principle, but my main point is that we should set a clear goal re what "standard JS" means and to what exact extent "common tools" (which?) should be able to process these packages.

Reply:
I don't think it needs to be that complicated. We should align with what libraries targeting web or authored in typescript are already doing. RN JS crashes when most tools try to parse it because it contains flow types in JS files. TS packages on npm don't have this problem because compiling `*.ts` to `*.js` + `*.d.ts` is part of the workflow for TS projects. So let's start by getting flow types out of the published JS files.

Reply:
> let's start by getting flow types out of the published JS files.

To be clear, I'm fully on board with this part. It's just that there's more related scope that's worth being clear about. (e.g. just removing Flow types doesn't mean we stop transpiling or speed up builds)

Reply:
Concretely, I would recommend that we narrow the wording here:

```suggestion
* Remove Flow types before publishing to npm (make this the default for the community). React Native and its community packages currently publish uncompiled, flow-typed JavaScript to npm. This causes common tools to crash on these packages.
```

Reply:
Many packages already use `react-native-builder-bob` for compiling code before publishing. So if there's a recommended configuration on how to compile, it'd be a good place to do so and will benefit a lot of libraries with a relatively simple way to migrate.

https://github.com/callstack/react-native-builder-bob